core/templates/dev/head/pages/exploration_player/PlayerServices.js

```diff
@@ -407,6 +422,28 @@ oppia.factory('oppiaPlayerService', [
           deferred.resolve(DEFAULT_PROFILE_IMAGE_PATH);
         }
         return deferred.promise;
+      },
+      recordSolutionHit: function() {
+        // Check that the current card is an active card by checking whether
+        // the active card's solution is the same as the current solution.
+        var activeStateName = playerPositionService.getCurrentStateName();
+        var activeCardSolution = exploration.getInteraction(
+          activeStateName).solution;
+        if (activeCardSolution == SolutionManagerService_getCurrentSolution()) {
+          StatsReportingService.recordSolutionHit(activeStateName);
+        }
+      },
+      recordStateFinished: function(activeCardIndex) {
+        var _editorPreviewMode = (
+          explorationContextService.getPageContext() === (
+            PAGE_CONTEXT.EDITOR));
+
+        if (!_editorPreviewMode && activeCardIndex - 1 >= 0) {
```

Comment:
Why the condition activeCardIndex - 1 >= 0? (And why not just pass stateName)?

Also, if the same state is encountered more than once in a playthrough, we consider it started N times and finished N times, right? (I.e. within a session it's fine to start a state multiple times and finish it multiple times?)

Reply:
@seanlip Is that so? How does the active card system work? 
Take this use case: If we went past card A and now in card B. An answer to card B sends us back to card A. At this point, which is the active card? A or B? I was in the mindset that the active card referred to the card that didn't have previous responses - so , in this case I was of the impression that card B would still be the active card. This way, we record StateFinish for each state only once - which is what we need.

Reply:
In your scenario, I'd have thought that A would be the active card. You can test this manually (and add console logs)...

Reply:
Redundant now. This function has been removed.